### PR TITLE
Updates to packaging scripts

### PR DESCRIPTION
**Pull request by M-DLB**
This PR contains enhancements to the Packaging scripts (CreateUCDComponentVersion and PackageBuildOutputs) to correctly calculate cumulative packages, when using multiple Build Reports. The key used in the Map object, used to accumulate artifacts to be deployed, is formed of the member name and the deploy type, instead of the fully-qualified dataset name and member in the previous versions.
It also contains an enhancement to the ArtifactRepositoryHelpers script, to support retries when performing uploads or downloads to/from an Artifact Repository server.

### CODE REVIEW

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy [hunk 1]
```diff
@@ -140,9 +182,9 @@ properties.buildReportOrder.each{ buildReportFile ->
 	deletions.each { deleteRecord ->
 		deletionCount += deleteRecord.getAttributeAsList("deletedBuildOutputs").size()
 		deleteRecord.getAttributeAsList("deletedBuildOutputs").each{ deletedFile ->
-			buildOutputsMap.put(deletedFile, [buildReport, deleteRecord])
+			def (dataset, member) = getDatasetName(deletedFile)
+			buildOutputsMap.put(new DeployableArtifact(member, "DELETE") , [dataset, buildReport, deleteRecord])
```
Comment (dennis-behm) [inline Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy:186]: Hi @M-DLB, I am running into an issue with this change to turn `member` + `deployType` into the key of the map: In my test scenario with multiple build reports, I have a scenario, that:
1. Adds a new member DELETEME
```
**  Deployable files detected in /var/jenkins/tmp/BuildReport-18.json
   JENKINS.ZDAT.RETIRE.LOAD(EBUD03), LOAD
   JENKINS.ZDAT.RETIRE.LOAD(EBUD02), LOAD
   JENKINS.ZDAT.RETIRE.LOAD(EBUD0RUN), LOAD
   JENKINS.ZDAT.RETIRE.LOAD(EBUD01), LOAD
   JENKINS.ZDAT.RETIRE.JCL(ALLOCDS), TRANSFER
   JENKINS.ZDAT.RETIRE.JCL(RUNZUNIT), TRANSFER
   JENKINS.ZDAT.RETIRE.JCL(EBUD0RUN), TRANSFER
   JENKINS.ZDAT.RETIRE.JCL(EBUD1RUN), TRANSFER
   JENKINS.ZDAT.RETIRE.JCL(DELETEME), TRANSFER
```
2. Deletes the member in a later stage in the process:
```
**  Deleted files detected in /var/jenkins/tmp/BuildReport-21.json
   JENKINS.ZDAT.RETIRE.SOURCE(DELETEME)
   JENKINS.ZDAT.RETIRE.JCL(DELETEME)
```

The new implementation carries now multiple entries: 
```
   Creating shiplist record for build output JENKINS.ZDAT.RETIRE.JCL(DELETEME) with recordType COPY_TO_PDS.
   Creating shiplist record for build output JENKINS.ZDAT.RETIRE.JCL(DELETEME) with recordType DELETE_RECORD.
   Defining shiplist delete container for JENKINS.ZDAT.RETIRE.JCL(DELETEME).
```
In this case UCD packaging fails with a duplicated member message:
```
** Create version by running UCD buztool
zOS toolkit config   : /var/ucd-agent/ (7.2.1.0,20211011-0758)
zOS toolkit binary   : /var/ucd-agent/ (7.2.1.0,20211011-0758)
zOS toolkit data set : RATCFG.UCD.V7R2M1 (7.2.1.0,20211011-0758)
Reading parameters:
....Command : createzosversion2
....Component : retirementCalculatorGithub
....Version : 169_20230424.093335.033
....Shiplist file : /var/jenkins/workspace/zunit-retirementCalculator/BUILD-169/build.20230424.113124.031/shiplist.xml
....Buztool Properties File : /var/jenkins/workspace/zunit-retirementCalculator//retirementCalculator/retirementCalculator/application-conf/retirementCalculator.ucd.properties
....Output File:/var/jenkins/workspace/zunit-retirementCalculator/BUILD-169/build.20230424.113124.031/buztool.output
Verifying version
....Repository location : /var/ucd-agent/var/repository/retirementCalculatorGithub/169_20230424.093335.033
Pre-processing shiplist:

*! Error executing buztool
Shiplist pre-processing error: duplicated member 'DELETEME' of JENKINS.ZDAT.RETIRE.JCL
```

I am attaching the full pipeline console output here - 
[consoleOutput-Duplicate-Member-Error.log](https://github.com/IBM/dbb/files/11319748/consoleOutput-Duplicate-Member-Error.log)

File: Pipeline/PackageBuildOutputs/README.md [hunk 2]
```diff
@@ -24,6 +24,9 @@ This section provides a more detailed explanation of how the PackageBuildOutputs
       1. Parse and extract build output information for records of type *ExecuteRecord* and *CopyToPDSRecord*. (Requires at least DBB 1.0.8.)
       1. Remove output entries that have no `deployType` set and remove unwanted outputs such as outputs with the `deployType` equal to `ZUNIT-TESTCASE`.
    1. If processing multiple build reports, a cumulative hashmap of output records is created to be able to combine outputs from multiple pipeline builds into a single tar file.
+   	  1. The key of the map, used in the calculation of the artifacts to be deployed, is the combination of the member name and the deploy type.
```
Comment (dennis-behm) [inline Pipeline/PackageBuildOutputs/README.md:27]: Can you please add a comment, that the script is generally not managing deletions? Something we missed to call out earlier.

**Review by dennis-behm — CHANGES_REQUESTED** [review]
@M-DLB, thank you for the enhancements to both scripts.  I understand the need to various HLQs.

As indicated in comment, the change for multiple build reports introduces a gap when a file is added and removed. We want to rethink about what is used as a key for the map when supporting multiple build reports (which applies to both scripts).

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy [hunk 3]
```diff
@@ -68,6 +68,41 @@ import groovy.json.JsonSlurper
  *  
  */
 
+class DeployableArtifact {
```
Comment (dennis-behm) [inline Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy:71]: Can we move the definition of the class to the bottom of the script and provide some inline documentation?

Comment (M-DLB) [reply Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy]: Done

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy [hunk 4]
```diff
@@ -160,8 +198,43 @@ properties.buildReportOrder.each{ buildReportFile ->
 			deletions.each { it.getAttributeAsList("deletedBuildOutputs").each { println("   ${it}")}}
 		}
 	}
+	buildReportRank++
 }
 
+
+println(tempBuildOutputsMap.keySet())
+
+// Remove duplicates
```
[the remark this answers is not in the document]
Comment (M-DLB) [reply Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy]: Done

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy [hunk 5]
```diff
@@ -160,8 +198,43 @@ properties.buildReportOrder.each{ buildReportFile ->
 			deletions.each { it.getAttributeAsList("deletedBuildOutputs").each { println("   ${it}")}}
 		}
 	}
+	buildReportRank++
 }
 
+
+println(tempBuildOutputsMap.keySet())
```
[the remark this answers is not in the document]
Comment (M-DLB) [reply Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy]: Done

File: Pipeline/PackageBuildOutputs/PackageBuildOutputs.groovy [hunk 6]
```diff
@@ -29,14 +31,53 @@ import groovy.cli.commons.*
...
- *  
+ *
+ * Version 5 - 2023-04
+ *  - Enhanced the multiple build reports feature to allow replacing the same artifact
+ *  if memberName and deployType are the same. The key in the HashMap is changed to
+ *  a DeployableArtifact object. 
+
  ************************************************************************************/
 
+class DeployableArtifact {
```
[the remark this answers is not in the document]
Comment (M-DLB) [reply Pipeline/PackageBuildOutputs/PackageBuildOutputs.groovy]: Done.

File: Pipeline/PackageBuildOutputs/PackageBuildOutputs.groovy [hunk 7]
```diff
@@ -123,106 +164,78 @@ props.buildReportOrder.each{ buildReportFile ->
...
+		if (props.addExtension && props.addExtension.toBoolean()) {
+			fileName = fileName + '.' + deployableArtifact.deployType
+		}
+		def file = new File(filePath, fileName)
+
+		// set copyMode based on last level qualifier
+		currentCopyMode = copyModeMap[dataset.replaceAll(/.*\.([^.]*)/, "\$1")]
+		if (currentCopyMode != null) {
+			// Copy outputs to HFS
+			CopyToHFS copy = new CopyToHFS()
```
[the remark this answers is not in the document]
Comment (M-DLB) [reply Pipeline/PackageBuildOutputs/PackageBuildOutputs.groovy:181]: I can add that check.

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy (hunk 1 above)
Comment (M-DLB) [reply Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy]: That should be now fixed. Can you check again with your testcase?

File: Pipeline/PackageBuildOutputs/README.md (hunk 2 above)
Comment (M-DLB) [reply Pipeline/PackageBuildOutputs/README.md:27]: Added a sentence about it.

File: Pipeline/PackageBuildOutputs/PackageBuildOutputs.groovy (hunk 7 above)
[the remark this answers is not in the document]
Comment (M-DLB) [reply Pipeline/PackageBuildOutputs/PackageBuildOutputs.groovy]: Check has been added:

```
** Copying build outputs to temporary package directory....
     Copying JENKINS.ZDAT.RETIRE.JCL(RUNZUNIT) to /u/gitlab/gitlab-runner/zos/builds/retirementCalculator/testDelete/BUILD-OUTPUT/tempPackageDir/JENKINS.ZDAT.RETIRE.JCL/RUNZUNIT.TRANSFER with DBB Copymode TEXT...
     Copying JENKINS.ZDAT.RETIRE.LOAD(EBUD02) to /u/gitlab/gitlab-runner/zos/builds/retirementCalculator/testDelete/BUILD-OUTPUT/tempPackageDir/JENKINS.ZDAT.RETIRE.LOAD/EBUD02.LOAD with DBB Copymode LOAD...
     Copying JENKINS.ZDAT.RETIRE.LOAD(EBUD0RUN) to /u/gitlab/gitlab-runner/zos/builds/retirementCalculator/testDelete/BUILD-OUTPUT/tempPackageDir/JENKINS.ZDAT.RETIRE.LOAD/EBUD0RUN.LOAD with DBB Copymode LOAD...
*! The file 'GITLAB.RETCALC.TESTDELE.LOAD(EBUD01)' doesn't exist. Copy is skipped.
     Copying JENKINS.ZDAT.RETIRE.JCL(EBUD0RUN) to /u/gitlab/gitlab-runner/zos/builds/retirementCalculator/testDelete/BUILD-OUTPUT/tempPackageDir/JENKINS.ZDAT.RETIRE.JCL/EBUD0RUN.TRANSFER with DBB Copymode TEXT...

```

**Review by dennis-behm — CHANGES_REQUESTED** [review]
Found one issue with multiple delete records - still continuing on the review.

See below comment

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy [hunk 8]
```diff
@@ -140,9 +146,9 @@ properties.buildReportOrder.each{ buildReportFile ->
 	deletions.each { deleteRecord ->
 		deletionCount += deleteRecord.getAttributeAsList("deletedBuildOutputs").size()
 		deleteRecord.getAttributeAsList("deletedBuildOutputs").each{ deletedFile ->
-			buildOutputsMap.put(deletedFile, [buildReport, deleteRecord])
+//			def (ds,member) = getDatasetName(deletedFile)
```
[the remark this answers is not in the document]
Comment (M-DLB) [reply Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy:146]: The problem was with the new JSON parser introduced with DBB 2.0.1. I've added a check to convert the object as a String and remove the extra double quotes (as it seems the GSON parser doesn't consider these entries in an array as Strings). It is compatible with DBB 1.1.x and DBB 2.0.1.
Can you please try again your testcase?

File: Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy [hunk 9]
```diff
@@ -329,13 +360,13 @@ xml.manifest(type:"MANIFEST_SHIPLIST"){
 			deletedFiles.each { deletedOutput ->
```
Comment (dennis-behm) [inline Pipeline/CreateUCDComponentVersion/dbb-ucd-packaging.groovy:360]: @M-DLB 

We need to clean the `deleteOutput` as well, like you do when you put it into the HashMap.
```
String cleansedDeletedFile = ((String) deletedFile).replace('"', '');
```

**Review by dennis-behm — APPROVED** [review]
I thoroughly tested the script updates with various BuildReports.  Thank you for the contribution!

**Pull request merged**